misc/python/materialize/cli/mzcompose.py

```diff
@@ -177,142 +189,251 @@ def gen_shortcuts(repo: mzbuild.Repository) -> int:
+
+    def run(cls, args: argparse.Namespace) -> None:
+        super().run(args)
+        repo = load_repo(args)
+        for name in sorted(repo.compositions):
+            print(name)
+
+
+class ListPortsCommand(Command):
+    name = "list-ports"
+    help = "list ports exposed by a service"
+
+    def configure(self, parser: argparse.ArgumentParser) -> None:
+        parser.add_argument(
+            "service", metavar="SERVICE", help="the service to list ports for"
         )
-    return 0
+
+    def run(self, args: argparse.Namespace) -> None:
+        super().run(args)
+        composition = load_composition(args)
+        for port in composition.find_host_ports(args.service):
+            print(port)
+
+
+class ListWorkflowsCommand(Command):
+    name = "list-workflows"
+    help = "list workflows in the composition"
+
+    def run(self, args: argparse.Namespace) -> None:
+        super().run(args)
+        composition = load_composition(args)
+        for name in sorted(
+            list(composition.yaml_workflows) + list(composition.python_funcs)
+        ):
+            print(name)
+
+
+class WebCommand(Command):
+    name = "web"
+    help = "open a service's URL in a web browser"
+
+    def configure(self, parser: argparse.ArgumentParser) -> None:
+        parser.add_argument("service", metavar="SERVICE", help="the service to target")
+
+    def run(self, args: argparse.Namespace) -> None:
+        super().run(args)
+        composition = load_composition(args)
+        ports = composition.find_host_ports(args.service)
+        if len(ports) == 1:
+            webbrowser.open(f"http://localhost:{ports[0]}")
+        elif not ports:
+            raise errors.MzRuntimeError(f"No running services matched {args.service!r}")
+        else:
+            raise errors.MzRuntimeError(
+                f"Too many ports matched {args.service!r}, found: {ports}"
+            )
 
 
-# We subclass `argparse.ArgumentParser` so that we can override its default
-# behavior of exiting on error. We want Docker Compose to be responsible for
-# generating option-parsing errors.
-class ArgumentParser(argparse.ArgumentParser):
-    def __init__(self) -> None:
-        super().__init__(add_help=False)
-        # If you add an option here, add it to the help text in `print_help`.
-        self.add_argument("--mz-quiet", action="store_true", default=None)
-        self.add_argument("--mz-find")
-        self.add_argument(
-            "--mz-build-mode", default="release", choices=["dev", "release"]
+class DockerComposeCommand(Command):
+    add_help = False
+
+    def __init__(self, name: str, help: str, acquire_deps: bool = False):
+        self.name = name
+        self.help = help
+        self.acquire_deps = acquire_deps
+
+    def configure(self, parser: argparse.ArgumentParser) -> None:
+        parser.add_argument("-h", "--help", action="store_true")
+
+    def run(self, args: argparse.Namespace) -> None:
+        if args.help:
```

Comment:
I think here we need to also add help for the workflow itself. It's out of scope for the pr, but the `mzworkflows.py` should be able to add help (like explaining how `TD_TEST`/positional version of `TD_TEST` works, and we should print that here, maybe after the `docker-compose` help. I like the `replace` below, but a log line referencing that "this subcommand will use docker-compose to run {workflow_name} workflow" or whatever would be really nice! otherwise, for new people it can be a bit inscrutable what is actually happening (even below we `announce` we are delegating)

Reply:
Yeah, that would that'd be awesome! `RunCommand` further overrides this help btw to note that it can run a workflow in addition to a service.